fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/MultiDigestInputStreamWrapper.java

```diff
@@ -0,0 +1,109 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.persistence.common;
+
+import static java.lang.String.format;
+import static org.apache.commons.codec.binary.Hex.encodeHexString;
+import static org.apache.commons.lang3.StringUtils.substringAfterLast;
+import static org.fcrepo.kernel.api.utils.ContentDigest.getAlgorithm;
+
+import java.io.InputStream;
+import java.net.URI;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.fcrepo.kernel.api.exception.InvalidChecksumException;
+import org.fcrepo.kernel.api.exception.RepositoryRuntimeException;
+
+/**
+ * Wrapper for an InputStream that allows for the computation and evaluation
+ * multiple digests at once
+ *
+ * @author bbpennel
+ */
+public class MultiDigestInputStreamWrapper {
+
+    private InputStream sourceStream;
+
+    private Map<String, String> algToDigest;
+
+    private Map<String, DigestInputStream> algToDigestStream;
+
+    /**
+     * Construct a MultiDigestInputStreamWrapper
+     *
+     * @param sourceStream the original source input stream
+     * @param digests collection of digests for the input stream
+     */
+    public MultiDigestInputStreamWrapper(final InputStream sourceStream, final Collection<URI> digests) {
+        this.sourceStream = sourceStream;
+        algToDigest = new HashMap<>();
+        algToDigestStream = new HashMap<>();
+
+        for (final URI digestUri : digests) {
+            final String algorithm = getAlgorithm(digestUri);
+            final String hash = substringAfterLast(digestUri.toString(), ":");
+            algToDigest.put(algorithm, hash);
+        }
+    }
+
+    /**
+     * Get the InputStream wrapped to produce the requested digests
+     *
+     * @return wrapped input stream
+     */
+    public InputStream getInputStream() {
+        InputStream digestStream = sourceStream;
+        for (final String algorithm : algToDigest.keySet()) {
+            try {
+                // Progressively wrap the original stream in layers of digest streams
+                digestStream = new DigestInputStream(
+                        digestStream, MessageDigest.getInstance(algorithm));
+            } catch (final NoSuchAlgorithmException e) {
+                throw new RepositoryRuntimeException(e);
+            }
+
+            algToDigestStream.put(algorithm, (DigestInputStream) digestStream);
+        }
+        return digestStream;
+    }
+
+    /**
+     * After consuming the inputstream, verify that all of the computed digests
+     * matched the provided digests.
+     *
+     * @throws InvalidChecksumException thrown if any of the digests did not match
+     */
+    public void checkFixity() throws InvalidChecksumException {
+        for (final var entry: algToDigestStream.entrySet()) {
```

Comment:
It seems like `checkFixity` relies on the code having called `getInputStream` first. If you call `checkFixity` without calling `getInputStream` `algToDigestStream` would have no entries so this doesn't check anything.

I think.

Reply:
that's correct, I tried to document this requirement in the javadoc for this method with "After consuming the inputstream"

Reply:
Maybe move the digest wrapping stuff to a private method that can be called if `algToDigestStream.size() != algToDigest.size()` ? Otherwise just use the existing `algToDigestStream`

Reply:
If you call checkFixity before getInputStream is called then you'd get a checksum mismatch because the digests it'd get back would be for a 0 byte stream, since nothing had been read so far. 

I could add an explicit failure if getInputStream hadn't been called yet, mainly as a warning to developers. It still wouldn't guarantee that the stream had been read though. I guess it could check that the stream was closed? Not 100% sure that's reliable, may depend on the type of inputstream...

Reply:
Okay, I added a check in to explicitly make sure getInputStream was called first